perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java

```diff
@@ -2718,6 +2720,208 @@ public void updateSponsorshipValidity(PerunSession sess, Member sponsoredMember,
 		return membersManagerImpl.getSponsorshipsExpiringInRange(sess, from, to);
 	}
 
+	@Override
+	public void moveMembership(PerunSession sess, Vo vo, User sourceUser, User targetUser) throws MemberNotExistsException, AlreadyMemberException, ExtendMembershipException {
+		Member sourceMember = this.getMemberByUserId(sess, vo, sourceUser.getId());
+		List<Group> directGroups = getPerunBl().getGroupsManagerBl().getMemberDirectGroups(sess, sourceMember);
+
+		List<Group> synchronizedGroups = new ArrayList<>();
+		try {
+			for (Group group : directGroups) {
+				Attribute attrSynchronizeEnabled = getPerunBl().getAttributesManagerBl().getAttribute(sess, group, getPerunBl().getGroupsManager().GROUPSYNCHROENABLED_ATTRNAME);
+				if ("true".equals(attrSynchronizeEnabled.getValue())) {
+					synchronizedGroups.add(group);
+				}
+			}
+		} catch (WrongAttributeAssignmentException | AttributeNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		Member targetMember;
+		try {
+			// create targetMember and add him to all direct groups that are not synchronized
+			directGroups.removeAll(synchronizedGroups);
+			targetMember = this.createMember(sess, vo, targetUser, directGroups);
+		} catch (WrongAttributeValueException | WrongReferenceAttributeValueException e) {
+			throw new InternalErrorException(e);
+		}
+
+		// move member related attributes
+		// add 'members' group to directGroups to move its member-group attributes too
+		try {
+			directGroups.add(getPerunBl().getGroupsManagerBl().getGroupByName(sess, vo, VosManager.MEMBERS_GROUP));
+		} catch (GroupNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+		moveMembersAttributes(sess, vo, sourceMember, targetMember, directGroups);
+
+		// set status from old member
+		// first validate member to prevent MemberNotValidYetException later
+		try {
+			this.validateMember(sess, targetMember);
+		} catch (WrongAttributeValueException | WrongReferenceAttributeValueException e) {
+			throw new InternalErrorException(e);
+		}
+		if (sourceMember.getStatus() != Status.VALID) {
+			try {
+				this.setStatus(sess, targetMember, sourceMember.getStatus());
+			} catch (WrongAttributeValueException | WrongReferenceAttributeValueException | MemberNotValidYetException e) {
+				throw new InternalErrorException(e);
+			}
+		}
+
+		// move all applications
+		getMembersManagerImpl().moveMembersApplications(sess, sourceMember, targetMember);
+
+		// move sponsorships in the VO
+		moveMembersSponsorships(sess, vo, sourceUser, targetUser, sourceMember, targetMember);
+
+		// move members bans
+		moveMembersBans(sess, sourceMember, targetMember);
+
+		// finally delete sourceMember
```

Comment:
This is a nice example, why these inline comments don't (usually) make sense. They are usually used in long methods (as in this example) and this actually hints, that one should split the method into multiple smaller methods. In general, the code itself should be self explonatary (with a help of javadoc). After you split it, these comments are just a copy of the methods' names. It is better to actually omit these comments since, after some time, they might diverge from the code, and after that, it might not be clear, whether the comment is invalid or the actual code is invalid. Again, you don't need to change it here, it is just a tip :wink: